### PR TITLE
Use CMake for Build System

### PR DESCRIPTION
**Pull request by WillAyd**
Proof of concept. I think we can greatly simplify our building system using this. @mroeschke @lithomas1 @Dr-Irv 

This is in an intermediate state, but you can do:

```sh
python setup.py
cmake .
make -j$(nproc)
```

`make clean` will clean up build artifacts, which right now excludes the generated Cython files but should include them in the future.

Have the following TODOs:
  [ ]: The CMAKE_SHARED_LIBRARY_SUFFIX is hard-coded to assume Py38 on linux
  [ ]: Get rid of `python setup.py` and have CMake perform cythonization
  [ ]: Update CI, documentation, etc...

### CODE REVIEW

**Comment by Dr-Irv** [comment]
I tried this on Windows.  Doing `python setup.py` caused the cythonize step to be done, but it was done serially.  Can't that be done in parallel as when we do `python setup.py build_ext -j 8` ?

Right now, cmake is not working, but I'm not sure I have the latest build tools installed.

**Comment by WillAyd** [comment]
Yea it could be. I didn't put a lot of time into that yet but if you want to explore would be great. I ideally want CMake to build the Cython extensions. That way running `make clean` should clean up those generated files, while we also would decouple a hard dependency on setuptools.

This is hard coded right now for Linux. For windows / your python version, you likely need to change like 8 of the top level CMake file to match the expected file ending for your platform:

`set(CMAKE_SHARED_LIBRARY_SUFFIX .cpython-38-x86_64-linux-gnu.so)`

This might be abstracted in some form via the CMake Python variables. If not something we can branch on in our file and upstream:

https://cmake.org/cmake/help/latest/module/FindPython.html#result-variables

**Comment by jbrockmendel** [comment]
> Well, I'm not sure I have the time nor understanding to get this working on Windows!

Yah, i expect this would be easier to get working on Mac, but it still involves a learning curve.  I'm fine embracing that for personal edification, curious how others feel

**Comment by WillAyd** [comment]
> Well, I'm not sure I have the time nor understanding to get this working on Windows!

Ah sorry...didn't think about make not working on Windows. I think you can do

```sh
cmake .
cmake --build .
```

On all systems to get this to work for an "in source" build

**Comment by WillAyd** [comment]
I think we can drop the minimum CMake version to 3.14 which would help. I just pulled the 3.22 off their documentation, but the NumPy finder was added in 3.14

For cleaning, you could ideally do `cmake --clean .` That will automatically get rid of files generated via build. In its current form that excludes Cython generated files, but in a future iteration it should do all of it

**Comment by Dr-Irv** [comment]
I made the following changes to `CMakeLists.txt`, and got a lot further on Windows:
```
cmake_minimum_required(VERSION 3.15)

set(CMAKE_SHARED_LIBRARY_SUFFIX .cp38-win_amd64.pyd)
```

Then I discovered that you have to specify the architecture to cmake.  So I did `cmake -A x64 --clean .`  I then got more adventurous and did a parallel build via `cmake  --build . --config Release --parallel 8` , discovering the parameters via some google searching.  The build succeeded, using all my cores, BUT the `pyd` files ended up in subdirectories called `Release` of `pandas/_libs/` and `pandas/_libs/tslibs/` etc.  

If I don't include `--config Release` it looks for a debugging version of the python library and fails.

Don't know ANYTHING about cmake to figure out how to get the files out of those Release subdirectories into the right place.

Also not sure that the `--clean` is removing previous files correctly (e.g., `.obj` files).

**Comment by WillAyd** [comment]
Awesome! I think you can tweak those settings via some CMAKE variables:

https://stackoverflow.com/questions/45368513/how-to-prevent-cmake-from-creating-debug-release-folders-under-executable-output

**Comment by WillAyd** [comment]
FYI the idiomatic approach to CMake is to do an "out of source" build, where we would have a dedicated build sub folder and all required elements would be copied into that. In such a case, I don't think the Release / Debug sub folders generated by VS would matter much.

However, we today do in source builds of pandas. So want to make sure we emulate that as best as possible for initial lift and shift

**Comment by Dr-Irv** [comment]
> However, we today do in source builds of pandas. So want to make sure we emulate that as best as possible for initial lift and shift

Once you figure that out for Linux, I hope it will be straightforward to do for Windows.  I'll wait to hear about your progress.

**Comment by WillAyd** [comment]
Sounds good. Thanks for the feedback so far - very helpful

**Comment by mroeschke** [comment]
I'd be supportive of trying CMake. Generally hoping

1. Not too much deviations across platforms Linux/MacOS/Windows
2. Near identical setup & invocation for local development & CI system

**Comment by WillAyd** [comment]
This should be closer to cross platform. Some notes since @Dr-Irv last tried.

1. The min required version is 3.17. This is because the Python_SOABI variable was introduce then, which gives us the required platform identifier
2. Python_SOABI seems to be missing the extension. I'm guessing this is pyd on Windows from what @Dr-Irv suggested before

Since the Cython files are generated during compilation, I think these should work in a cross-platform manner:

```sh
cmake .
cmake --build . --parallel
cmake --build . --target clean
```

**Comment by Dr-Irv** [comment]
> 1. The min required version is 3.17 (released March 2020). This is because the Python_SOABI variable was introduced then, which gives us the required platform identifier

I think that may require an update to Visual Studio 2022 build tools on Windows, and right now the pandas development docs say to use VS 2019, which has cmake 3.15.  So if we are to use this on Windows, we'd have to do one of the following:

- figure out how to update `cmake` in a VS 2019 installation and document it accordingly
- move to VS 2022 and see if pandas can be built using the newer version.

I'm not going to be experimenting with that!!!

File: pandas/_libs/src/ujson/python/ujson.c [hunk 1]
```diff
@@ -74,7 +74,7 @@ static PyModuleDef moduledef = {
 };
 
 
-PyMODINIT_FUNC PyInit_json(void) {
+PyMODINIT_FUNC PyInit_ujson(void) {
```
Comment (WillAyd) [inline pandas/_libs/src/ujson/python/ujson.c:77]: This was clobbering the built in json module used by cythonize

**Comment by jbrockmendel** [comment]
IIUC scipy recentlyish changed to use meson.  Is the ecosystem converging on any build patterns?

**Comment by WillAyd** [comment]
Good question to which I do not know the answer. My only reason for picking CMake was familiarity with it through some Arrow work

**Comment by eli-schwartz** [comment]
SciPy wrote a blog post about moving to Meson: https://labs.quansight.org/blog/2021/07/moving-scipy-to-meson/

It has some useful links and discussion (note that a year later and with various collaboration between SciPy <--> Meson, not everything described there is still the same, in particular we added native Cython support, fixed the prefix issue, and SciPy now uses https://pypi.org/project/meson-python/ as a native build backend to run).

In the comments, there's also a link discussing a proposal to improve the build backend that uses cmake (scikit-build).

Whichever build system you choose, you do NOT want to have to manually glue setuptools and something else together.

**Comment by lithomas1** [comment]
> IIUC scipy recentlyish changed to use meson. Is the ecosystem converging on any build patterns?

I'm pretty sure numpy wants to use meson as well. (cc @rgommers). 

I'm a little worried about CMake since I have no experience with it, and I am currently the only other person who maintains the wheel builds other than @simonjayhawkins.

Whatever we do, a clean out-of-the box build of pandas should just work, and there should be no extra dependencies not installed from PyPI(other than the C Compiler).

**Comment by rgommers** [comment]
> I'm pretty sure numpy wants to use meson as well. (cc @rgommers).

Yes indeed. NumPy and scikit-learn are planning to move to Meson, following SciPy, and I think other projects in the ecosystem are going to follow that lead. The SciPy 1.9.0 release (close to ready, RC1 hopefully this week) is the first release to default to Meson.

As for CMake, it is clearly a much more capable build system than distutils/setuptools, so it would be an improvement to move to it. For SciPy, CMake was the only other build system we seriously considered besides Meson. There's discussion on the choice in https://github.com/scipy/scipy/issues/13615. To summarize the main reasons:
- (a) Meson is written in Python so easier to contribute to,
- (b) Meson has much better documentation (although there's good third-party CMake docs/books), and 
- (c) Meson is overall a more well-designed and easier to use build system. This is an opinion clearly and not everyone will agree, but I think it has been confirmed by SciPy devs who were only familiar with CMake beforehand, and found Meson more pleasant to touch. If CMake would remove everything that is legacy or "not best-practice", it would probably be similar to Meson in usability and user-friendliness. The trouble is that the CMake devs won't do that, and the sum total of CMake is a bit of a mess.

I'll also add, now that the SciPy migration is mostly done, that the Meson devs have been quite helpful and responsive to questions and issues (they implemented Cython support for example, and improved usability around virtualenvs).

A few thoughts on this PR and the proposed Pandas change:
- I see that the diff here deletes `setup.py`. I recommend not doing that, but having two build systems in parallel for one release until you know everything works (because it won't - niche platforms and more niche use cases are hard to test for example).
- What Pandas needs from a build system is not that much, so it's not that difficult to experiment. Either CMake or Meson should have all you need _build-wise_.
- The more important topic here is what you need _Python packaging-wise_. The `pyproject.toml` (PEP 517) hook mechanism is immature, and it was a big job to build and debug [meson-python](https://github.com/FFY00/meson-python). Neither CMake nor Meson know about the details of Python packaging or `pyproject.toml`, so you need that kind of a "shim package". For CMake, that is [scikit-build](https://github.com/scikit-build/scikit-build). It still heavily depends on `setuptools` and needs an overhaul. That's not as much of a problem for Pandas as it would have been for SciPy or NumPy, because latest `setuptools` is incompatible with `numpy.distutils` - which you don't need in Pandas. That said, I would investigate that, as @eli-schwartz also pointed out above.
- From my perspective it would be preferable synergy-wise if all packages near the core of the PyData ecosystem made the same build system choice.
  - That said, there is also something to say for having two viable build system choices for packages with compiled code. It's more effort overall, but in the long run it could be healthy.
  - I'll note that large projects that use (or plan to use) CMake, like PyTorch or RAPIDS, do typically not distribute sdists on PyPI (or don't support PyPI at all in the RAPIDS case), so they don't have to worry about the `pyproject.toml` / PEP 517 support that's still lacking as of today.
- I'll also note that support-wise, there is a NASA grant across NumPy, Pandas, SciPy and scikit-learn which was recently awarded (we'll need to write a blog post about that next week to share the details). This grant covers 0.5 FTE for 3 years specifically for build & packaging work. For that work it would be great if these four packages used the same build system. That said, if Pandas is going to make a different choice, I'm sure we can find some CMake expertise to help support that choice if needed (significant work on overhauling scikit-build is out of scope though).
- In the end what is most important is that the maintainers who do the heavy lifting on build & packaging issues are happy with the final choice.

**Comment by WillAyd** [comment]
Thanks @eli-schwartz and @rgommers for the feedback- lots of great consideration points. I'll give meson a look along with this. Sounds like Meson is great for the packages you've pointed out and Cmake gets us closer to Arrow (@jorisvandenbossche may have insight for that). 

Either choice should represent a nice upgrade over our current system

File: Dockerfile [hunk 2]
```diff
@@ -2,7 +2,7 @@ FROM quay.io/condaforge/miniforge3
 
 # if you forked pandas, you can pass in your own GitHub username to use your fork
 # i.e. gh_username=myname
-ARG gh_username=pandas-dev
+ARG gh_username=WillAyd
```
Comment (WillAyd) [inline Dockerfile:5]: This is temporary. There is a gap with this Dockerfile currently where it clones the main repo, but obviously none of the CMake stuff is in main

**Comment by Dr-Irv** [comment]
I installed cmake into my conda environment on Windows.  removed all cmakefiles from previous trials.  Then pulled down the PR, and got this:
```
cmake --build . --target clean
Error: could not load cache
```

**Comment by WillAyd** [comment]
@Dr-Irv did you run `cmake .` first? That should configure everything and create CMakeCache.txt

**Comment by WillAyd** [comment]
> CMake Error: CMAKE_GENERATOR was set but the specified generator doesn't exist. Using CMake default.

I'm guessing this might be the error?

> Then left out the target, and got a bunch of errors related to not finding `python38_d.lib`

IIUC VS builds things differently. On Unix platforms you specify release / debug at configure time. MSVC will generate release and debug simultaneously at build time if you don't specify `--config Released

You might need to clean and try again for the remaining errors. My guess is some of the debug / release profiles generated by VS got mixed up

**Comment by Dr-Irv** [comment]
> Yea I think the different build types got mixed up when running the build without specifying `--config Release` though

I'm still stuck with those errors about `numpy.core.multiarray failed to import`

**Comment by WillAyd** [comment]
Definitely on the tip of this branch right? There were some previous issues where we would build an internal json module and then `import json` From numpy or Cython would pick that up instead of the stdlib. Seems similar to the traceback you have. Worth double checking there is no json.so in your pandas._libs directory

**Comment by Dr-Irv** [comment]
> Definitely on the tip of this branch right? There were some previous issues where we would build an internal json module and then `import json` From numpy or Cython would pick that up instead of the stdlib. Seems similar to the traceback you have. Worth double checking there is no json.dll in your pandas._libs directory

Yes, there was a `json.pyd` file hanging around.  I deleted all the `pyd` files and built it and it worked!

Important that for Windows, you need to do `cmake . build --config Release` for it to work so that should be reflected in the docs.

Also, some files are left behind, and that could lead to someone committing them by mistake. The first 3 are there for each thing that is compiled.  
```
        pandas/_libs/tslibs/timedeltas.dir/
        pandas/_libs/tslibs/timedeltas.vcxproj
        pandas/_libs/tslibs/timedeltas.vcxproj.filters
        pandas.sln
        pandas/_libs/Release/
        pandas/_libs/tslibs/Release/
        pandas/_libs/window/Release/
        pandas/io/sas/Release/
        ALL_BUILD.vcxproj
        ALL_BUILD.vcxproj.filters
        ZERO_CHECK.vcxproj
        ZERO_CHECK.vcxproj.filters
        x64/
```

**Comment by WillAyd** [comment]
Awesome! Those `Release` folders shouldn't be there - my guess is they are from a previous instance you ran before we changed some of the cmake settings. When running `cmake clean` it will remove any files that it generated, but if you start with an earlier version of these changes cmake might no longer recongize that it once created those `Release` folders (this is the same issue you had with the json file)

For the .sln and vcxproj files I'm guessing we can gitignore them (?)

**Comment by WillAyd** [comment]
> Important that for Windows, you need to do `cmake . build --config Release` for it to work so that should be reflected in the docs.

Should be there in the contributing sections. Maybe we should add a note to highlight it more clearly though?

I didn't put it where it didn't matter, like in the debugging extensions .rst file since that doesn't currently provide windows instructions. Note that on Windows you would do:

```sh
cmake .
cmake --build . --config Release --parallel
```

Whereas other platforms you choose between Debug / Release at configure time (default is release)

```sh
cmake . -DCMAKE_BUILD_TYPE=Release    # can also exclude this flag for same result
cmake --build . --parallel      # can also invoke make directly
```

I've abstracted the nuances here in the docs by just showing the Windows instructions shown above, as it gets you ultimately to the same place on all platforms

**Comment by Dr-Irv** [comment]
> > Important that for Windows, you need to do `cmake . build --config Release` for it to work so that should be reflected in the docs.
> 
> Should be there in the contributing sections. Maybe we should add a note to highlight it more clearly though?

I don't think so.  I didn't look at the doc changes until now, and they are clear.

**Comment by WillAyd** [comment]
> I definitely removed the `Release` folders before the build. They are being left behind. Each folder has a `.exp` and `.lib` file for each thing that was compiled. Here's an example:

Hmm strange. I thought the `set(CMAKE_LIBRARY_OUTPUT_DIRECTORY_RELEASE "")` line would disable this. If you still get pandas to run fine after the build it might not be a bad thing. Let's see what we find

File: pandas/tests/groupby/test_groupby.py [hunk 3]
```diff
@@ -2552,7 +2550,6 @@ def test_groupby_numerical_stability_sum_mean(func, values):
     tm.assert_frame_equal(result, expected)
 
 
-@pytest.mark.xfail(not IS64, reason="GH#38778: fail on 32-bit system")
```
Comment (jbrockmendel) [inline pandas/tests/groupby/test_groupby.py:2555]: what fixed this?  is it possible to do independently of cmake?

Comment (WillAyd) [reply pandas/tests/groupby/test_groupby.py:2555]: Didn’t dive too deep into the 32 but stuff. I’m guessing something with the compilation flags would have done this. Here we enforce C99 standard behavior explicitly and whatever CMake chooses to send as flags to the compiler doesn’t necessarily match what distutils might have chosen

**Comment by jreback** [comment]
scipy and numpy are working on https://github.com/mesonbuild/meson

is there a reason you are pushing for cmake?

**Comment by WillAyd** [comment]
> scipy and numpy are working on https://github.com/mesonbuild/meson
> 
> is there a reason you are pushing for cmake?

cmake is used by Arrow and scikit-learn, so its not without its own precedent. AFAICT meson doesn't support in source builds, so it would also be a larger shift for us currently to move towards that compared to cmake. 

If someone wants to champion meson over cmake in the future I don't think this work is throwaway as it helps decouple us from setup tools hackery and gets us a lot closer to true PEP-517 support

**Comment by eli-schwartz** [comment]
@rgommers

> * If CMake would remove everything that is legacy or "not best-practice", it would probably be similar to Meson in usability and user-friendliness. The trouble is that the CMake devs won't do that, and the sum total of CMake is a bit of a mess.

(This cannot solve cmake's lack of e.g. a proper array type, or indeed, types. I don't know if it's theoretically possible, even, because that is not exactly a backwards-compatible file format / syntax change.)

@WillAyd

> cmake is used by Arrow and scikit-learn, so its not without its own precedent. AFAICT meson doesn't support in source builds, so it would also be a larger shift for us currently to move towards that compared to cmake.

In-source builds are problematic TBH, cmake recommends you do out of source builds for exactly this reason. It doesn't forbid in-source builds, but many projects then do so manually using strequal comparisons to error out if an in-source build is detected.

This actually kind of sucks because at that point cmake has already created some in-source files that confuses cmake when it tries to set up an out of source build after that. :(

Apparently that's already been problematic in this PR, witness stuff like:

> Also, some files are left behind, and that could lead to someone committing them by mistake. The first 3 are there for each thing that is compiled.

> When running `cmake clean` it will remove any files that it generated, but if you start with an earlier version of these changes then reconfigured cmake with newer settings at a later point in time, cmake might no longer recongize that it once created those `Release` folders

Out of source builds means that you can reliably nuke the entire build tree at once, it's one directory to `rm -rf`. You should use them -- even with cmake, even if the build system doesn't inherently forbid them.

(Also Meson sets up the build tree so that you don't accidentally commit those files to git. This is only possible because it's always out of source.)

> If someone wants to champion meson over cmake in the future I don't think this work is throwaway as it helps decouple us from setup tools hackery and gets us a lot closer to true PEP-517 support

It's a bit difficult to make out everything that this PR changes, because the commits don't distinguish between different kinds of WIP incremental changes. But based on the overall diff it seems like there isn't much in this PR that affects the project code, it's almost exclusively setuptools / cmake changes.

Assuming that setuptools is kept and used as build backend glue, this PR can be useful in transitioning to another (non-cmake) build system, as it decouples setuptools-glue from "build the artifacts", a useful goal.

But if someone champions Meson, then Meson *has* a true PEP-517 build backend, and doesn't use setuptools at all, not even as glue, so there's no real benefit to first refactoring it, then deleting it.

**Comment by WillAyd** [comment]
Yea agreed an out of source build is definitely preferable, but I think we'd have to do a lot more updates to our CI with that assumption as well as change developer expectations. Hoping the latter wouldn't be much of a bottleneck, but the former would definitely make this PR larger

**Comment by rgommers** [comment]
> cmake is used by Arrow and scikit-learn,

Not the latter - scikit-learn uses setuptools, and is most likely to move to Meson (work is planned to start in the coming month). scikit-image and others scikits I can think off all still use either setuptools or numpy.distutils

**Comment by WillAyd** [comment]
Ah thanks for clarifying. I thought the scikit-build package was associated with scikit-learn

**Comment by jbrockmendel** [comment]
> I think we can greatly simplify our building system using this

What's the criterion for judging this?  Is it just LOC?

I think there's an implicit "avoiding setuptools is desirable" consensus that I'm out of the loop on.

**Comment by eli-schwartz** [comment]
> I think there's an implicit "avoiding setuptools is desirable" consensus that I'm out of the loop on.

Setuptools invokes the compiler in an ad-hoc fashion, it's difficult to parallelize and *very* difficult to cross-compile. Those are among the bigger reasons SciPy switched from distutils to Meson.

**Comment by WillAyd** [comment]
To help clarify some more I think this is also worth drafting up via the new PDEP process @datapythonista is working on. This PR is part research, discussion and implementation at this point, so probably a good fit for that new process

**Comment by lithomas1** [comment]
> To help clarify some more I think this is also worth drafting up via the new PDEP process @datapythonista is working on. This PR is part research, discussion and implementation at this point, so probably a good fit for that new process

Yeah, I think it is worth pursing a PDEP. There's a lot of insight in the comments of this PR that I don't want to get lost, and are helpful when someone wants to revisit the build system or figure out rationale behind certain decisions. GitHub tends to handle PRs with lots of comments/activity super horribly, and kind of randomly collapses comments that could be important in the middle.

I think it is also worth pursuing a parallel PR using meson as the build backend. I am willing to do that if no one beats me to it.

**Comment by rgommers** [comment]
> I think there's an implicit "avoiding setuptools is desirable" consensus that I'm out of the loop on.

@jbrockmendel for a bit more context:
- There is indeed a consensus of people working on build stuff for scientific packages (and beyond) that `distutils` and everything that is related (`setuptools`, `numpy.distutils`, etc.) is very badly designed and basically unfixable. Tools like Cython and Pythran just all monkeypatch the internals to add the functionality they need, nothing is extensible. The `setuptools` devs are not going to fix this - and realistically, they can't.
- There have been multiple attempts before at moving away from `setuptools`, for example NumPy and SciPy had two working alternative build systems based on Scons and then Waf. The problem was always that tools like Pip assumed setuptools, and only more recently it's become possible to really move away via PEP 517, PEP 518, PEP 621 & co.
- The planned removal of `distutils` from Python 3.12 set the timer for a bunch of stuff; `numpy.distutils` is not compatible with the vendored `setuptools.distutils`, so we are in a hurry to move - about a year left (max, much faster would be better to not block other packages working with 3.12-dev). _That said, Pandas is not in that position, because it works fine with latest `setuptools`, so you can do things at your leisure._

For new projects, the only situation where you'd still want to prefer `setuptools` is for packages that do contain a few C/Cython extensions, but need almost nothing from a build system (cross-compiling, good performance, etc.). For pure Python packages there are much better options like Flit, or perhaps Poetry/Hatch/PDM (there's a long list, because it's now fairly easy to write such a thing for pure Python code). And if you do have serious needs from a build system - which includes Pandas and all packages at the center of the PyData ecosystem I'd say - then I'd say you want either Meson or CMake.

**Comment by lithomas1** [comment]
Hello everyone, I haven't really commented much the progress on using Meson as pandas' build system, but I now can compile basically all the c extensions. I'm still working on enabling some more extensions and cleaning up a lot of the boilerplate code, but I will try opening a draft PR next week. 

The next step would be to build wheels/figure out how to install locally. I'm still working through some gnarly issues in meson-python for that. 

All in all, we should probably start thinking of drafting up the PDEP sometime soon.

**Comment by simonjayhawkins** [comment]
> The next step would be to build wheels

can you confirm we would still be migrating to cibuildwheel #44027 to build wheels? https://github.com/pandas-dev/pandas/issues/44027#issuecomment-1076952117

**Comment by lithomas1** [comment]
> > The next step would be to build wheels
> 
> can you confirm we would still be migrating to cibuildwheel #44027 to build wheels? [#44027 (comment)](https://github.com/pandas-dev/pandas/issues/44027#issuecomment-1076952117)

Yes, that would be independent of this. Both build backends(Cmake, meson) are invoked through PEP 517 hooks.

> > The next step would be to build wheels
> 
> can you confirm we would still be migrating to cibuildwheel #44027 to build wheels? [#44027 (comment)](https://github.com/pandas-dev/pandas/issues/44027#issuecomment-1076952117)

Both meson and this Cmake implementation use PEP 517 build hooks, so cibuildwheel would build the wheels the same way(through [build](https://pypa-build.readthedocs.io/en/stable/)).

**Comment by lithomas1** [comment]
My WIP meson work can be found here lithomas1/pandas#19.

File: MANIFEST.in [hunk 4]
```diff
@@ -28,7 +30,6 @@ global-exclude *.odt
 global-exclude *.orc
 global-exclude *.sas7bdat
 global-exclude *.sav
-global-exclude *.so
```
Comment (WillAyd) [inline MANIFEST.in:31]: Right now this allows the .so to make it into the wheel build as part of setuptools, but unfortunately also includes as in the sdist. Open item that needs to be addressed but figured this was easiest to do for now to continue CI

**Comment by WillAyd** [comment]
@Dr-Irv any insight on the typing failure in CI? Does this stem from pandas-stubs having an entry for pandas._libs.json? Not sure how those should be synced

Otherwise I think this PR is in a usable state. Started on PDEP but lost a little motivation, and this was pretty close so decided to finish this instead

**Comment by Dr-Irv** [comment]
> @Dr-Irv any insight on the typing failure in CI? Does this stem from pandas-stubs having an entry for pandas._libs.json? Not sure how those should be synced
> 
I'm hazarding a guess here, but I think that you didn't include any of the `.pyi` files in the build.  This is not from the `pandas-stubs` project.  Rather, in `pandas/_libs` (and underneath that), there are multiple `.pyi` files that should be included in the distribution.

**Comment by WillAyd** [comment]
Thanks @Dr-Irv makes sense. Was overthinking that one

This is all green now if its something we'd still entertain. I think beyond this there's a lot we can do to improve our build system (moving to out of source builds comes to mind first and foremost) but I think this represents a reasonable lift and shift

**Comment by lithomas1** [comment]
Hey @WillAyd, I'm happy to help you draft up the PDEP if you need it. 
It'd probably be best to start off on something like a Google doc, as I expect there will be quite a few revisions needed before its ready for review as a formal PDEP.
I just discovered the pandas-core mailing list, so you can find my email through there if you need me.

**Comment by WillAyd** [comment]
OK sure just created https://github.com/pandas-dev/pandas/pull/47988 for that

**Comment by WillAyd** [comment]
After discussion seems like Meson is the way to go. Let's close this one

**Pull request closed without merging**